File: 4-backend/analytics-py-service/app/middleware/trust_kong.py
from fastapi import Request
from typing import Optional
import logging

logger = logging.getLogger(__name__)

class UserContext:
    """用户上下文，从Kong传递的headers中提取"""
    def __init__(self, request: Request):
        self.id = request.headers.get("x-user-id")
        self.role = request.headers.get("x-user-role")
        self.name = request.headers.get("x-user-name")
        self.email = request.headers.get("x-user-email")
    
    @property
    def is_authenticated(self) -> bool:
        return self.id is not None
    
    def __str__(self):
        return f"User(id={self.id}, role={self.role}, name={self.name})"

async def trust_kong_middleware(request: Request, call_next):
    """
    Kong信任中间件
    从Kong传递的headers中提取用户信息，无条件信任
    """
    # 提取用户信息
    user_context = UserContext(request)
    
    # 将用户信息添加到request state
    request.state.user = user_context
    
    # 记录请求日志
    if user_context.is_authenticated:
        logger.info(f"Request from {user_context} to {request.url.path}")
    else:
        logger.info(f"Anonymous request to {request.url.path}")
    
    # 继续处理请求
    response = await call_next(request)
    
    return response

def get_current_user(request: Request) -> Optional[UserContext]:
    """
    获取当前用户信息
    """
    return getattr(request.state, 'user', None)

def require_user(request: Request) -> UserContext:
    """
    要求用户认证（通过Kong）
    """
    user = get_current_user(request)
    if not user or not user.is_authenticated:
        from fastapi import HTTPException
        raise HTTPException(
            status_code=401,
            detail="Authentication required - No user info from Kong"
        )
    return user

def require_role(request: Request, required_roles: list) -> UserContext:
    """
    要求特定角色权限
    """
    user = require_user(request)
    if user.role not in required_roles:
        from fastapi import HTTPException
        raise HTTPException(
            status_code=403,
            detail=f"Insufficient permissions. Required roles: {required_roles}"
        )
    return user

File: 3-nodejs-middle-platform/src/middleware/trust-kong.ts
import { Request, Response, NextFunction } from 'express';

// Kong传递的用户信息头部
interface KongUserHeaders {
  'x-user-id': string;
  'x-user-role': string;
  'x-user-name': string;
  'x-user-email': string;
}

declare global {
  namespace Express {
    interface Request {
      user?: {
        id: string;
        role: string;
        name: string;
        email: string;
      };
    }
  }
}

/**
 * Kong信任中间件
 * 从Kong传递的Headers中提取用户信息，无条件信任
 */
export const trustKong = (req: Request, res: Response, next: NextFunction) => {
  // 从Kong传递的headers中提取用户信息
  const userId = req.headers['x-user-id'] as string;
  const userRole = req.headers['x-user-role'] as string;
  const userName = req.headers['x-user-name'] as string;
  const userEmail = req.headers['x-user-email'] as string;

  // 如果有用户信息，设置到request对象
  if (userId) {
    req.user = {
      id: userId,
      role: userRole,
      name: userName,
      email: userEmail
    };
  }

  next();
};

/**
 * 检查用户是否已认证（通过Kong传递的headers）
 */
export const requireKongAuth = (req: Request, res: Response, next: NextFunction) => {
  if (!req.user || !req.user.id) {
    return res.status(401).json({
      success: false,
      message: 'Authentication required - No user info from Kong'
    });
  }
  next();
};

/**
 * 检查用户角色权限
 */
export const requireRole = (roles: string[]) => {
  return (req: Request, res: Response, next: NextFunction) => {
    if (!req.user || !req.user.role) {
      return res.status(401).json({
        success: false,
        message: 'Authentication required'
      });
    }

    if (!roles.includes(req.user.role)) {
      return res.status(403).json({
        success: false,
        message: 'Insufficient permissions'
      });
    }

    next();
  };
};

/**
 * 日志记录中间件 - 记录用户操作
 */
export const logUserAction = (action: string, module: string) => {
  return (req: Request, res: Response, next: NextFunction) => {
    if (req.user) {
      // 这里可以记录用户操作日志
      console.log(`User ${req.user.id} performed ${action} in ${module} module`);
    }
    next();
  };
};